azurelinuxagent/ga/signature_validation.py

```diff
@@ -18,22 +18,82 @@
 #
 import base64
 import os
+import re
 
 from azurelinuxagent.common import conf
 from azurelinuxagent.common.utils.shellutil import run_command, CommandError
-from azurelinuxagent.common.exception import ExtensionError
-from azurelinuxagent.common import event
-from azurelinuxagent.common.event import WALAEventOperation
-from azurelinuxagent.common.exception import ExtensionErrorCodes
+from azurelinuxagent.common.exception import AgentError
+from azurelinuxagent.common import logger
+from azurelinuxagent.common.event import WALAEventOperation, add_event
 from azurelinuxagent.ga.signing_certificate_util import get_microsoft_signing_certificate_path
+from azurelinuxagent.common.utils.flexible_version import FlexibleVersion
 
 
-class SignatureValidationError(ExtensionError):
+# This file tracks the state of signature validation for the package. If the file exists, signature has been validated.
+_SIGNATURE_VALIDATION_STATE_FILE = "package_signature_and_manifest_validated"
+
+# Signature validation requires OpenSSL version 1.1.0 or later. The 'no_check_time' flag used for the 'openssl cms -verify'
+# command is not supported on older versions.
+_MIN_OPENSSL_VERSION_FOR_SIG_VALIDATION = FlexibleVersion("1.1.0")
+
+
+class PackageValidationError(AgentError):
+    """
+    Error raised when validation fails for a package.
+    """
+    def __init__(self, msg=None, inner=None, code=-1):
+        super(PackageValidationError, self).__init__(msg, inner)
+        self.code = code
+
+
+class SignatureValidationError(PackageValidationError):
     """
     Error raised when signature validation fails for an extension.
     """
 
 
+class ManifestValidationError(PackageValidationError):
+    """
+    Error raised when handler manifest 'signingInfo' validation fails for an extension.
```

Comment:
Looks like the signature validation code will be common between agent/extension. 
```suggestion
    Error raised when handler manifest 'signingInfo' validation fails for a package.
```

Reply:
Updated